sdk/cosmos/src/authorization_policy.rs

```diff
@@ -0,0 +1,320 @@
+                "Authorization policies cannot be the last policy of a pipeline".to_owned(),
+            )));
+        }
+
+        let time = chrono::Utc::now().format(TIME_FORMAT).to_string();
+
+        let uri_path = &request.uri().path_and_query().unwrap().to_string()[1..];
+        trace!("uri_path used by AuthorizationPolicy == {:#?}", uri_path);
+
+        let auth = {
+            let resource_link = generate_resource_link(&uri_path);
+            debug!("resource_link == {}", resource_link);
+            generate_authorization(
+                &self.authorization_token,
+                &request.method(),
+                &ctx.get_bag().resource_type,
+                resource_link,
+                &time,
+            )
+        };
+
+        trace!(
+            "AuthorizationPolicy calculated authorization == {}: {}",
+            AUTHORIZATION,
+            &auth
+        );
+
+        request
+            .headers_mut()
+            .append(HEADER_DATE, HeaderValue::from_str(&time)?);
+        request
+            .headers_mut()
+            .append(HEADER_VERSION, HeaderValue::from_static(AZURE_VERSION));
+        request
+            .headers_mut()
+            .append(AUTHORIZATION, HeaderValue::from_str(&auth)?);
+
+        trace!("\n\nrequest =={:?}", request);
+
+        // now next[0] is safe (will not panic) because we checked
+        // at the beginning of the function.
+        next[0].send(ctx, request, &next[1..]).await
+    }
+}
+
+// TODO: will become private as soon as cosmos_client will be migrated
+// to pipeline arch.
+pub(crate) fn generate_resource_link(u: &str) -> &str {
+    static ENDING_STRINGS: &[&str] = &[
+        "dbs",
+        "colls",
+        "docs",
+        "sprocs",
+        "users",
+        "permissions",
+        "attachments",
+        "pkranges",
+        "udfs",
+        "triggers",
+    ];
+
+    // store the element only if it does not end with dbs, colls or docs
+    let p = u;
+    let len = p.len();
+    for str_to_match in ENDING_STRINGS {
+        let end_len = str_to_match.len();
+
+        if end_len <= len {
+            let end_offset = len - end_len;
+            let sm = &p[end_offset..];
+            if sm == *str_to_match {
+                if len == end_len {
+                    return "";
+                }
+
+                if &p[end_offset - 1..end_offset] == "/" {
+                    let ret = &p[0..len - end_len - 1];
+                    return ret;
+                }
+            }
+        }
+    }
+    p
+}
+
+// TODO: make it private after pipeline migration
+pub(crate) fn generate_authorization(
+    auth_token: &AuthorizationToken,
+    http_method: &http::Method,
+    resource_type: &ResourceType,
+    resource_link: &str,
+    time: &str,
+) -> String {
+    let string_to_sign = string_to_sign(http_method, resource_type, resource_link, time);
+    debug!(
+        "generate_authorization::string_to_sign == {:?}",
```

Comment:
I'm not so sure how helpful this is. If you think it's useful, perhaps using `trace!` would be better?